challenges/systems/challenge-2.md

```diff
@@ -0,0 +1,296 @@
+  ensure you have enough time to work on the implementation.
+* Avoid writing an overly detailed design document. Two to three pages is
+  sufficient.
+* Avoid sending us draft design documents. It is difficult to evaluate which
+  parts are draft and which parts are complete. Instead we encourage asking
+  questions in Slack and sharing a design document that is ready to be reviewed.
+
+Once the design document has been approved by two reviewers, move on to the
+implementation.
+
+Split your code submission into roughly 3-5 Pull Requests to give the team an
+opportunity to review your code and provide feedback. Feel free to merge each
+PR after you have two approvals.
+
+Our team will do their best to provide a high quality review of the submitted
+Pull Requests in a reasonable time frame. You are spending your time on this,
+we are going to contribute our time too.
+
+After the final submission, the interview team will assemble and vote using a
+"+1, -2" anonymous voting system: +1 is submitted whenever a team member
+accepts the submission, -2 otherwise.
+
+In case of a positive result, we will connect you to our HR and recruiting
+teams, who will work out the details and present an offer.
+
+In case of a negative score result, the hiring manager will contact you and share a
+list of the key observations from the team that affected the result.
+
+# Requirements
+
+## Level 1
+
+### Library
+
+* Implement a request forwarder to forward requests to an upstream.
+
+### Server
+
+* Secure communication between client and server using TLS and an [opaque
+  bearer token](https://latacora.micro.blog/2018/06/12/a-childs-garden.html)
+  for authentication
+* Develop a simple authorization scheme that defines what clients can use the
+  load balancer; this scheme can be statically defined in code
+* Accept and forward requests to upstreams using the library
+
+## Level 2
+
+### Library
+
+* Implement a round-robin request forwarder to forward requests to
+  multiple upstreams.
+
+### Server
+
+* Use mTLS authentication to verify the client's identity
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Accept and forward requests to upstreams using library
+
+## Level 3
+
+### Library
+
+* Implement a least connections request forwarder that tracks the number of
+  connections per upstream.
+
+### Server
+
+* Use mTLS authentication to have the server verify identity of the client and
+  client of the server
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Accept and forward requests to upstreams using library
+
+## Level 4
+
+### Library
+
+* Implement a least connections request forwarder that tracks the number of
+  connections per upstream.
+* Implement a per-client connection rate limiter that tracks the number of
+  client connections.
```

Comment:
We've had a few applicants asking whether the ask is to "limit the number of concurrent client connections" or "limit the number of client connections within some period of time". Is this an intended gotcha, or can we improve the wording to not point towards the incorrect interpretations. (Not sure if I should be more explicit since this is public to interviewees...)

Reply:
True, but the phrasing "tracks the number of client connections" implies the other approach. IMO simply `Implement a per-client connection rate limiter` would be sufficient.